### PR TITLE
Accurate glyph metrics using a Glyph's Bounding Box

### PR DESCRIPTION
**Pull request by Silverwolf90**
This is a WIP.

Note that this is one of the changes outlined in https://github.com/0xfe/vexflow/issues/349

* Add a `BoundingBoxComputation` class -- [which was completely taken from here](https://github.com/icons8/svg-path-bounding-box/blob/master/lib/Path/BoundingBox.js)
  * This is only needed for a temporary amount of time. With SMuFL metadata we would no longer need a utilitiy to determine glyph bounding boxes.
* Add static method `Glyph.getOutlineBoundingBox`
  * Did a bit of refactoring to share logic between this and `Glyph.renderOutline`
* Every time `glyph.reset()` is called, the `bbox` property is set
* `Glyph.getMetrics()` will now return the `width` and `height` of the `bbox`
* Removed `Glyph.setWidth()` since there should be no reason to manually set the width of a glyph
  * This was only used in `Vex.Flow.KeySignature` -- that call has been removed

@0xfe The regression tests don't run on Windows. Any other options?

### CODE REVIEW

**Comment by Silverwolf90** [comment]
Screenshot of some bounding boxes:

image

File: src/keysignature.js [hunk 1]
```diff
@@ -108,7 +108,6 @@ Vex.Flow.KeySignature = (function() {
       var glyph_width = glyph_data.width + extra_width;
       this.width += glyph_width;
       // Set the width and place the glyph on the stave
-      glyph.setWidth(glyph_width);
```
Comment (0xfe) [inline src/keysignature.js]: Given that `glyph_width` here includes `extra_width`, do you know how it affects the final rendering of key signatures?

**Comment by 0xfe** [comment]
Thanks. Looks great overall. I'll run the visual tests and verify.

(BTW, Travis build seems to have failed.)

File: src/glyph.js [hunk 2]
```diff
@@ -35,12 +35,12 @@ Vex.Flow.Glyph = (function() {
...
     };
 
-    this.width = null;
     this.metrics = null;
     this.x_shift = 0;
     this.y_shift = 0;
 
-    if (options) this.setOptions(options); else this.reset();
+    if (options) this.setOptions(options);
+    else this.reset();
```
Comment (0xfe) [inline src/glyph.js]: here be dragons... can you brace these if/else blocks?

File: src/keysignature.js (hunk 1 above)
Comment (0xfe) [inline src/keysignature.js]: Okay, never mind, this is fine.

**Comment by gristow** [comment]
How would you feel about modifying this (or my modifying this) to also extract the y value(s) for the min & max x points, and vice-versa? I'm thinking that in the case of stem-to-notehead connections, we need both x & y for max & min x to make an accurate connection. (The change should be easy for the bezier & quadratic calculations, since these are finding x,y for the vertical & horizontal local maxima points already. We'd need to save last x and y for line-to to be able to make this calculation.)

(Potentially this could later be extended further to implement true collision detection, which would allow us to, for instance, space accidentals not in vertical columns, but nudge them in as close to the notes as possible, which would bring our accidental layouts 100% in line with Gould's _Behind Bars_.)

**Comment by 0xfe** [comment]
The build seems to have passed. Is this ready for review, or are you still working on it?

**Comment by Silverwolf90** [comment]
@gristow the min/max y values can be derived from the `Vex.Flow.BoundingBox` that gets set to `glyph.bbox` when `glyph.reset()` is called. Of course this could be exposed more directly through an API.

@0xfe No luck with cygwin yet (to be fair I gave up rather quickly, working with Linux day-to-day has ruined my Windows development experience...). The build passing is good enough for me to say it's ready.

**Comment by 0xfe** [comment]
Cyril, can you sync and fix the conflicts (from @aaronmars' recent PR)? I'll run the regression tests locally and merge.

**Comment by Silverwolf90** [comment]
Yeah I'll take care of that tonight!

**Comment by Silverwolf90** [comment]
good to go

**Comment by 0xfe** [comment]
Seems like a whole bunch of tests are failing in `grunt test`:

```
>> TextNote - TextNote Formatting With Glyphs 1
>> Message: Died on test #1     at test (file:///Users/mmuthanna/git/vexflow/tests/support/qunit.js:1222)
>>     at file:///Users/mmuthanna/git/vexflow/build/vexflow-tests.js:124
>>     at file:///Users/mmuthanna/git/vexflow/build/vexflow-tests.js:66
>>     at file:///Users/mmuthanna/git/vexflow/build/vexflow-tests.js:12660
>>     at file:///Users/mmuthanna/git/vexflow/build/vexflow-tests.js:14535
>>     at file:///Users/mmuthanna/git/vexflow/tests/flow.html:26
>>     at file:///Users/mmuthanna/git/vexflow/tests/support/jquery.js:3099
>>     at file:///Users/mmuthanna/git/vexflow/tests/support/jquery.js:3211
>>     at file:///Users/mmuthanna/git/vexflow/tests/support/jquery.js:3417
>>     at completed (file:///Users/mmuthanna/git/vexflow/tests/support/jquery.js:3433): 'undefined' is not a function (evaluating 'bboxComp.addPoint.bind(bboxComp)')
>> Actual: null
>> Expected: undefined
>> TypeError: 'undefined' is not a function (evaluating 'bboxComp.addPoint.bind(bboxComp)')
```

**Comment by 0xfe** [comment]
Okay, that was because you updated qunit, and I didn't `npm install`.

**Comment by 0xfe** [comment]
Looks like a lot of tests were affected by this. I'll send you a link to the images in a sec. Here are the scores meanwhile (highest is worst, i.e., most difference between blessed and current):

```
StaveModifier.Begin___End_StaveModifier_Test 104.458
Stave.Multiple_Stave_Repeats_Test 20.3731
Annotation.Harmonics 12.5387
KeySignature.Cancelled_key_test 12.1431
KeySignature.Stave_Helper 6.80729
Clef_Keys.Minor_Key_Clef_Test 2.41326
Clef_Keys.Major_Key_Clef_Test 2.41326
Beam.Mixed_Beam_1 1.95491
KeySignature.Minor_Key_Test 0.62087
KeySignature.Major_Key_Test 0.62087
Curve.Rounded_Curve 0.609138
Clef_Keys.Stave_Helper 0.505903
StaveNote.Center_Aligned_Note_with_Multiple_Modifiers 0.26101
Tuplet.Mixed_Stem_Direction_Bottom_Tuplet 0.225196
KeySignature.Altered_key_test 0.192622
Beam.Break_Secondary_Beams 0.172665
Three_Voice_Rests.Auto_Adjust_Rest_Positions___Three_Voices__1 0.170972
Bend.Whako_Bend 0.114862
Bend.Bend_Phrase 0.113681
Bend.Reverse_Bends 0.110996
Curve.Thick_Thin_Curves 0.0896378
Bend.Double_Bends 0.0793729
Accidental.Automatic_Accidentals___C_major_scale_in_Ab 0.0479428
Accidental.Automatic_Accidentals___No_Accidentals_Necsesary 0.0295291
Accidental.Automatic_Accidentals___Multi_Voice_Inline 0.0259743
Bend.Double_Bends_With_Release 0.0205035
TimeSignature.Big_Signature_Test 0.0203115
Stave.Multiple_Staves_Volta_Test 0.0202506
Accidental.Automatic_Accidentals___Multi_Voice_Offset 0.0164436
StaveConnector.Bold_Double_Lines_Offset_Draw_Test_2 0.013174
Tuplet.Bottom_Ratioed_Tuplet 0.0131318
StaveNote.StaveNote_Draw___Alto 9.87733e-05
Auto_Beaming.Flat_Beams_Up_Bounds 9.80282e-05
Tuplet.Bottom_Tuplet 9.61748e-06
Percussion.Percussion_Snare1 9.51987e-06
Annotation.Bottom_Annotations_with_Beams 9.49276e-05
Percussion.Percussion_Basic1 9.02637e-05
Rests.Auto_Align_Rests___Beamed_Notes_Stems_Down 8.76991e-06
PedalMarking.Custom_Text_1 8.65824e-05
PedalMarking.Simple_Pedal_1 8.65506e-05
PedalMarking.Simple_Pedal_3 8.64229e-05
PedalMarking.Release_and_Depress_on_Same_Note_1 8.6289e-05
PedalMarking.Simple_Pedal_2 8.62454e-05
PedalMarking.Custom_Text_2 8.61811e-05
PedalMarking.Release_and_Depress_on_Same_Note_2 8.6176e-05
Beam.Sixteenth_Beam 8.38005e-05
Three_Voice_Rests.Three_Voices____2_Complex 8.14598e-05
StaveNote.StaveNote_Draw___Bass 8.03635e-05
StringNumber.String_Number_In_Notation 7.95476e-05
Annotation.Test_Justification_Annotation_Stem_Up 7.20831e-06
Three_Voice_Rests.Three_Voices____3 7.20116e-05
Rests.Auto_Align_Rests___Tuplets_Stems_Down 7.15733e-05
Auto_Beaming.Beam_Across_All_Rests_with_Stemlets 6.84494e-05
TabSlide.Simple_TabSlide 6.76042e-05
Beam.TabNote_Beams_Up 6.61782e-09
Beam.Lengthy_Beam 5.8813e-06
Auto_Beaming.Break_Beams_on_Rest 5.35152e-06
Auto_Beaming.Flat_Beams_Mixed_Direction 4.73575e-06
Percussion.Percussion_Basic0 4.59041e-06
StringNumber.Complex_Measure_With_String___Finger_Numbers 4.56816e-06
Auto_Beaming.Flat_Beams_Down 4.38132e-05
TabTie.Simple_TabTie 4.32364e-07
Beam.TabNote_Beams_Down 4.23362e-08
TextNote.TextNote_Formatting_With_Glyphs_1 4.0784e-05
Auto_Beaming.More_Simple_Auto_Beaming_0 4.04131e-05
Formatter.Format_Multiple_Staves___Justified 4.01074e-05
TextBracket.Simple_TextBracket 3.87902e-05
StaveNote.StaveNote_BoundingBoxes___Treble 3.25212e-05
Tuplet.Awkward_Tuplet 3.2189e-06
Auto_Beaming.Custom_Beam_Groups 3.21859e-05
Tuplet.Simple_Tuplet 3.17136e-05
Tuplet.Complex_Tuplet 3.15287e-05
Rhythm.Rhythm_Draw___slash_notes 2.69015e-06
Beam.Simple_Beam 2.68869e-05
Beam.Multi_Beam 2.4959e-05
NoteHead.Basic 2.42012e-06
Percussion.Percussion_Basic2 2.41428e-05
Vibrato.Simple_Vibrato 2.41163e-06
TimeSignature.Time_Signature_multiple_staves_alignment_test 2.35194e-07
Rhythm.Rhythm_Draw___beamed_slash_notes__some_rests 2.29692e-05
Annotation.Test_Justification_Annotation_Stem_Down 2.21487e-05
Vibrato.Harsh_Vibrato 1.96283e-06
Beam.Close_Trade_offs_Beam 1.92527e-05
StaveTie.No_Start_Note 1.92399e-05
Auto_Beaming.More_Simple_Tuplet_Auto_Beaming 1.90607e-06
Auto_Beaming.Break_Beams_on_Middle_Rests_Only 1.76129e-05
StaveLine.Simple_StaveLine 1.73192e-05
Voice.Full_Voice_Mode_Test 1.56501e-05
StaveLine.StaveLine_Arrow_Options 1.54599e-05
Percussion.Percussion_Snare2 1.50987e-05
Annotation.Simple_Annotation 1.42732e-05
Annotation.Bottom_Annotation 1.42085e-05
Rests.Rests___Dotted 1.34684e-05
Auto_Beaming.Flat_Beams_Down__uniform_ 1.29372e-05
Beam.Slopey_Beam 1.27254e-05
TabTie.Continuous 1.19869e-05
TextBracket.TextBracket_Styles 1.04226e-05
TabTie.Tapping 0.00782715
TimeSignature.Time_Signature_Change_Test 0.0075004
StaveConnector.Bold_Double_Lines_Offset_Draw_Test 0.00744117
Vibrato.Vibrato_with_Bend 0.00625538
Auto_Beaming.Even_Group_Stem_Directions 0.00590914
Annotation.Fingerpicking 0.00530619
Auto_Beaming.Beat_group_with_unbeamable_note___2_2 0.00529396
TextNote.TextNote_Formatting_With_Glyphs_0 0.00477966
StringNumber.Fret_Hand_Finger_In_Notation 0.00432768
Beam.Mixed_Beam_2 0.00413137
Auto_Beaming.Odd_Group_Stem_Directions 0.00357569
Beam.Complex_Beams_with_Articulations 0.00337588
StaveNote.Center_Aligned_Note_with_Articulation 0.0029362
StaveNote.Center_Aligned_Note 0.0029362
Rhythm.Rhythm_Draw___32nd_note_rhythm_with_scratches 0.00273944
Auto_Beaming.Odd_Time___Guessing_Default_Beam_Groups 0.0027183
Auto_Beaming.More_Automatic_Beaming 0.00264361
TabSlide.Slide_Down 0.00240439
Rhythm.Rhythm_Draw___16th_note_rhythm_with_scratches 0.00227191
Three_Voice_Rests.Auto_Adjust_Rest_Positions___Three_Voices__2 0.00224655
Auto_Beaming.Maintain_Stem_Directions 0.00217633
Formatter.Notes_with_Tab 0.00217169
Auto_Beaming.More_Simple_Auto_Beaming_1 0.00202553
Rhythm.Rhythm_Draw___beamed_slash_notes 0.00194265
Clef.Clef_Change_Test 0.00186801
Auto_Beaming.Maintain_Stem_Directions___Beam_Over_Rests 0.00172132
Grace_Notes.Grace_Notes_Multiple_Voices 0.00166125
Accidental.Automatic_Accidentals 0.00145217
Auto_Beaming.Duration_Based_Secondary_Beam_Breaks 0.00130418
Stave.Tempo_Test 0.00112783
Clef.Clef_End_Test 0.00109206
Annotation.Standard_Notation_Annotation 0.00107727
Auto_Beaming.Simple_Auto_Beaming 0.00101005
Curve.Top_Curve 0.000973901
TabSlide.Slide_Up 0.000748622
Beam.Auto_stemmed_Beam 0.000661099
Formatter.StaveNote_Justification 0.000633057
StaveNote.Center_Aligned_Note___Multi_Voice 0.000585616
Three_Voice_Rests.Three_Voices____1 0.000525776
Formatter.Format_Multiple_Staves___No_Justification 0.000515865
Clef.Small_Clef_End_Test 0.000497208
Grace_Notes.Grace_Note_Basic_with_Slurs 0.000475062
Beam.Outlier_Beam 0.000457939
Stave.Multiple_Stave_Barline_Test 0.000426535
Curve.Simple_Curve 0.000424483
Auto_Beaming.Odd_Beam_Groups_Auto_Beaming 0.000407355
StaveNote.Center_Aligned_Note_with_Annotation 0.000401324
Clef.Clef_Test 0.000383101
Tuplet.Beamed_Tuplet 0.000372016
Beam.Complex_Beams_with_Annotations 0.000368638
TabTie.Pulloffs 0.000350838
TabTie.Hammerons 0.000350834
Auto_Beaming.Offset_beat_grouping___6_8_ 0.000340589
Tuplet.Mixed_Stem_Direction_Tuplet 0.000314165
Percussion.Percussion_Snare0 0.000288381
TimeSignature.Basic_Time_Signatures 0.000287791
Strokes.Strokes___Multi_Voice_Notation_and_Tab 0.000256591
Tuplet.Ratioed_Tuplet 0.000246231
Auto_Beaming.Flat_Beams_Down_Bounds 0.000243248
StaveNote.StaveNote_Draw___Treble 0.000241865
Auto_Beaming.Simple_Tuplet_Auto_Beaming 0.000226173
Rests.Auto_Align_Rests___Beamed_Notes_Stems_Up 0.000219958
Beam.Dotted_Beam 0.000212311
Strokes.Strokes___Notation_and_Tab 0.000199694
Grace_Notes.Grace_Note_Basic 0.000181078
TextNote.Crescendo 0.000173262
Auto_Beaming.Flat_Beams_Up 0.00017214
Rests.Auto_Align_Rests___Single_Voice__Align_All_ 0.000159174
Auto_Beaming.Beam_Across_All_Rests 0.000150072
NoteHead.Bounding_Boxes 0.000137178
Clef.Small_Clef_Test 0.000132719
StaveNote.StaveNote_Draw___Tenor 0.000131772
Rests.Auto_Align_Rests___Tuplets_Stems_Up 0.000130785
Rests.Auto_Align_Rests___Multi_Voice 0.000130144
StaveNote.StaveNote_Draw___Bass_2 0.000129547
Rests.Auto_Align_Rests___Single_Voice__Default_ 0.000121348
Tuplet.Nested_Tuplets 0.000120012
Beam.Insane_Beam 0.000111613
Percussion.Percussion_Notes 0.000111584
Auto_Beaming.Flat_Beams_Up__uniform_ 0.000106393
Three_Voice_Rests.Auto_Adjust_Rest_Positions___Two_Voices 0
TextNote.Text_Dynamics 0
TextNote.TextNote_Superscript_and_Subscript 0
TextNote.TextNote_Formatting 0
TabStave.Vertical_Bar_Test 0
TabStave.TabStave_Draw_Test 0
TabNote.TabNote_Stems_with_Dots 0
TabNote.TabNote_Stems_Up_Through_Stave 0
TabNote.TabNote_Stems_Up 0
TabNote.TabNote_Stems_Down_Through_Stave 0
TabNote.TabNote_Stems_Down 0
TabNote.TabNote_Draw 0
Strokes.Strokes___Multi_Voice 0
Strokes.Strokes___Brush_Arpeggiate_Rasquedo 0
StringNumber.Multi_Voice_With_Strokes__String___Finger_Numbers 0
StaveTie.Stem_Up_StaveTie 0
StaveTie.Simple_StaveTie 0
StaveTie.Set_Direction_Up 0
StaveTie.Set_Direction_Down 0
StaveTie.No_End_Note 0
StaveTie.Chord_StaveTie 0
StaveNote.StaveNote_Draw___StaveNote_Styles 0
StaveNote.StaveNote_Draw___Slash 0
StaveNote.StaveNote_Draw___Key_Styles 0
StaveNote.StaveNote_Draw___Harmonic_And_Muted 0
StaveNote.Flag_and_Dots_Placement___Stem_Down 0
StaveNote.Flag_and_Dot_Placement___Stem_Up 0
StaveNote.Displacements 0
StaveNote.Beam_and_Dot_Placement___Stem_Up 0
StaveNote.Beam_and_Dot_Placement___Stem_Down 0
StaveModifier.Vertical_Bar_Test 0
StaveModifier.Stave_Draw_Test 0
StaveHairpin.Vertical_Offset_StaveHairpin 0
StaveHairpin.Simple_StaveHairpin 0
StaveHairpin.Horizontal_Offset_StaveHairpin 0
StaveHairpin.Height_StaveHairpin 0
StaveConnector.Thin_Double_Line_Right_Draw_Test 0
StaveConnector.Single_Draw_Test__1px_Stave_Line_Thickness 0
StaveConnector.Single_Draw_Test 0
StaveConnector.Single_Both_Sides_Test 0
StaveConnector.Double_Draw_Test 0
StaveConnector.Combined_Draw_Test 0
StaveConnector.Bracket_Draw_Test 0
StaveConnector.Brace_Wide_Draw_Test 0
StaveConnector.Brace_Draw_Test 0
StaveConnector.Bold_Double_Lines_Overlapping_Draw_Test 0
StaveConnector.Bold_Double_Line_Right_Draw_Test 0
StaveConnector.Bold_Double_Line_Left_Draw_Test 0
Stave.Vertical_Bar_Test 0
Stave.Stave_Text_Test 0
Stave.Stave_Draw_Test 0
Stave.Single_Line_Configuration_Test 0
Stave.Open_Stave_Draw_Test 0
Stave.Multiple_Line_Stave_Text_Test__Raphael_ 0
Stave.Batch_Line_Configuration_Test 0
Percussion.Percussion_Clef 0
Ornament.With_Upper_Lower_Accidentals 0
Ornament.Stacked 0
Ornament.Ornaments___Delayed_turns 0
Ornament.Ornaments_Vertically_Shifted 0
Ornament.Ornaments 0
Formatter.StaveNote_Formatting 0
Dot.Multi_Voice 0
Dot.Basic 0
Beam.TabNote_Beams_Auto_Stem 0
Beam.TabNote_Auto_Create_Beams 0
Barline.Simple_BarNotes 0
Articulation.TabNote_Articulation 0
Articulation.Articulation___Up_stroke_Down_Stroke 0
Articulation.Articulation___Staccato_Staccatissimo 0
Articulation.Articulation___Snap_Pizzicato_Fermata 0
Articulation.Articulation___Marcato_L_H__Pizzicato 0
Articulation.Articulation___Inline_Multiple 0
Articulation.Articulation___Fermata_Above_Below 0
Articulation.Articulation___Accent_Tenuto 0
Annotation.TabNote_Annotations 0
Accidental.Stem_Down 0
Accidental.Multi_Voice 0
Accidental.Microtonal 0
Accidental.Basic 0
Accidental.Accidental_Arrangement_Special_Cases 0
```

**Comment by 0xfe** [comment]
Here are the failing tests -- blessed, current, diff for each test.

tuplet mixed_stem_direction_bottom_tuplet_blessed
tuplet mixed_stem_direction_bottom_tuplet_current
tuplet mixed_stem_direction_bottom_tuplet
tuplet bottom_ratioed_tuplet_blessed
tuplet bottom_ratioed_tuplet_current
tuplet bottom_ratioed_tuplet
timesignature big_signature_test_blessed
timesignature big_signature_test_current
timesignature big_signature_test
three_voice_rests auto_adjust_rest_positions___three_voices__1_blessed
three_voice_rests auto_adjust_rest_positions___three_voices__1_current
three_voice_rests auto_adjust_rest_positions___three_voices__1
stavenote center_aligned_note_with_multiple_modifiers_blessed
stavenote center_aligned_note_with_multiple_modifiers_current
stavenote center_aligned_note_with_multiple_modifiers
stavemodifier begin___end_stavemodifier_test_blessed
stavemodifier begin___end_stavemodifier_test_current
stavemodifier begin___end_stavemodifier_test
staveconnector bold_double_lines_offset_draw_test_2_blessed
staveconnector bold_double_lines_offset_draw_test_2_current
staveconnector bold_double_lines_offset_draw_test_2
stave multiple_staves_volta_test_blessed
stave multiple_staves_volta_test_current
stave multiple_staves_volta_test
stave multiple_stave_repeats_test_blessed
stave multiple_stave_repeats_test_current
stave multiple_stave_repeats_test
keysignature major_key_test_blessed
keysignature major_key_test_current
keysignature major_key_test
keysignature minor_key_test_blessed
keysignature minor_key_test_current
keysignature minor_key_test
keysignature stave_helper_blessed
keysignature stave_helper_current
keysignature stave_helper
keysignature altered_key_test_blessed
keysignature altered_key_test_current
keysignature altered_key_test
keysignature cancelled_key_test_blessed
keysignature cancelled_key_test_current
keysignature cancelled_key_test
clef_keys stave_helper_blessed
clef_keys stave_helper_current
clef_keys stave_helper
curve rounded_curve_blessed
curve rounded_curve_current
curve rounded_curve
curve thick_thin_curves_blessed
curve thick_thin_curves_current
curve thick_thin_curves
clef_keys minor_key_clef_test_blessed
clef_keys minor_key_clef_test_current
clef_keys minor_key_clef_test
clef_keys major_key_clef_test_blessed
clef_keys major_key_clef_test_current
clef_keys major_key_clef_test
bend whako_bend_blessed
bend whako_bend_current
bend whako_bend
bend double_bends_with_release_blessed
bend double_bends_with_release_current
bend double_bends_with_release
bend reverse_bends_blessed
bend reverse_bends_current
bend reverse_bends
bend double_bends_blessed
bend double_bends_current
bend double_bends
bend bend_phrase_blessed
bend bend_phrase_current
bend bend_phrase
beam mixed_beam_1_blessed
beam mixed_beam_1_current
beam mixed_beam_1
beam break_secondary_beams_blessed
beam break_secondary_beams_current
beam break_secondary_beams
annotation harmonics_blessed
annotation harmonics_current
annotation harmonics
accidental automatic_accidentals___multi_voice_inline_blessed
accidental automatic_accidentals___multi_voice_inline_current
accidental automatic_accidentals___multi_voice_inline
accidental automatic_accidentals___multi_voice_offset_blessed
accidental automatic_accidentals___multi_voice_offset_current
accidental automatic_accidentals___multi_voice_offset
accidental automatic_accidentals___no_accidentals_necsesary_blessed
accidental automatic_accidentals___no_accidentals_necsesary_current
accidental automatic_accidentals___no_accidentals_necsesary
accidental automatic_accidentals___c_major_scale_in_ab_blessed
accidental automatic_accidentals___c_major_scale_in_ab_current
accidental automatic_accidentals___c_major_scale_in_ab

**Comment by 0xfe** [comment]
Most of them seem okay, just micro adjustments that don't matter a lot. Some of the key sig accidantals look a bit tight.

**Comment by Silverwolf90** [comment]
Ah, found the source of the `KeySignature` problems, it's the only class that was using the `glyph.setWidth()` override. So I added a commit that preserves previous formatting in an array called `keySignature.xPositions` (which is relative to the `keySignature.x`) instead of overriding each `glyph.width`.

**Comment by Silverwolf90** [comment]
Something to take into consideration is what we consider "public," if it includes **every** instance method on each class then technically this PR would be a breaking change since I removed the `glyph.setWidth()` method.

**Comment by SalahAdDin** [comment]
:+1:

**Comment by 0xfe** [comment]
Good point re: `qlyph.setWidth()`. I don't really consider that public because that whole class is an internal detail to VexFlow, but maybe we should just keep the function and emit a deprecated warning on the console when its called?

**Comment by 0xfe** [comment]
Also, I see `keySignature` is fixed. Let me know when this is good to go and I'll run the regression tests again.

**Comment by Silverwolf90** [comment]
If you don't consider it public (and I don't either), I'd prefer the more aggressive approach and simply remove it. I don't think it's sustainable to deprecate all questionable functionality. Particularly without a designated timeline for complete removal. With a slow moving project like VexFlow, leaving in bad functionality can only contribute to code rot (which the codebase already suffers from pretty extensively). The more aggressive change likely breaks a very small minority users, who are probably power-users if they've gone that deep into the internals.

**Comment by Silverwolf90** [comment]
The regression tests are ready to go.

**Comment by 0xfe** [comment]
Okay, yeah, I can get behind that.

**Comment by 0xfe** [comment]
Regression tests look good. Merging!

**Pull request merged**